### PR TITLE
Implement agg(distinct) handler

### PR DESCRIPTION
**Pull request by marqueurs404**

### CODE REVIEW

File: djongo/sql2mongo/functions.py [hunk 1]
```diff
@@ -19,12 +20,22 @@ def token2sql(token: Token,
                   query: 'query_module.BaseQuery'
                   ) -> U['CountFunc',
                          'SimpleFunc']:
+        ## FIX: agg(distinct)
+        if isinstance(token[0], Identifier) and token[1][1].match(tokens.Keyword, 'DISTINCT'):
+            token.is_agg_distinct = True
+        else:
+            token.is_agg_distinct = False
```
Comment (morbitech1) [inline djongo/sql2mongo/functions.py]: if/else redundant:
token.is_agg_distinct = isinstance(token[0], Identifier) and token[1][1].match(tokens.Keyword, 'DISTINCT')

File: djongo/sql2mongo/converters.py [hunk 2]
```diff
@@ -77,31 +78,52 @@ def to_mongo(self):
...
+                if isinstance(selected, SQLFunc):
+                    ## FIX: issue occurs when there's no explicit alias and we're dealing with FROM(subquery)
+                    alias = selected.alias or str(selected.__hash__())
+                    group[alias] = selected.to_mongo()
+                    project[alias] = True
+                else:
+                    project[selected.field] = True
+            final_pipeline = [{'$group': group}, {'$project': project}]
+
+        return final_pipeline
```
Comment (morbitech1) [inline djongo/sql2mongo/converters.py]: This section is a bit messy (a lot of nested if/else) Also you are repeating the code for  the loop.

I would prefer if you could refactor some of the code that runs in the loop into some smaller functions, and reuse the for loop logic?

**Comment by morbitech1** [comment]
@marqueurs404 I've added some code to pass all the migrations, feel free to review

File: djongo/sql2mongo/converters.py (hunk 2 above)
Comment (marqueurs404) [reply djongo/sql2mongo/converters.py]: i've done some clean up, let me know if you have any ideas on making it better

File: djongo/sql2mongo/converters.py [hunk 3]
```diff
@@ -77,32 +78,62 @@ def to_mongo(self):
...
+                # token = agg(col)
+                key = self._get_alias(selected)
+                group[key] = selected.to_mongo()
+                pipeline = [{'$group': group}]
+            else: # token = col
+                key = selected.field
+                pipeline = [{'$group': group}]
+            project[key] = True
+            keys.append(key)
+            pipelines.append(pipeline)
```
Comment (morbitech1) [inline djongo/sql2mongo/converters.py]: I would turn the code running inside the for loop here into a function, in that way you can turn the pipelines, keys into a list comprehension:
pipelines, keys = zip(*[self._handle_agg_distinct_token(...) for selected in self.sql_tokens])

File: djongo/sql2mongo/converters.py [hunk 4]
```diff
@@ -77,32 +78,62 @@ def to_mongo(self):
...
+
+        if len(pipelines) > 1: 
+            # use $facet when there are multiple selected columns
+            facet, project = {}, {}
+            for idx, pipeline in enumerate(pipelines):
+                facet[str(idx)] = pipeline
+                project[keys[idx]] = {'$arrayElemAt': [ f"${idx}.{keys[idx]}", 0]}
+            final_pipeline = [{'$facet': facet}, {'$project': project}]
+        else:
+            final_pipeline = pipelines[0] + [{'$project': project}]
```
Comment (morbitech1) [inline djongo/sql2mongo/converters.py]: This code would look a lot cleaner if you just do the negative case first:
```
if len(pipelines) == 1:
    return pipelines[0] + [{'$project': project}]
# use $facet when there are multiple selected columns
facet, project = {}, {}
for idx, pipeline in enumerate(pipelines):
    facet[str(idx)] = pipeline
    project[keys[idx]] = {'$arrayElemAt': [ f"${idx}.{keys[idx]}", 0]}
return [{'$facet': facet}, {'$project': project}]
```

**Comment by marqueurs404** [comment]
> @marqueurs404 I've added some code to pass all the migrations, feel free to review

Hmm I seem to be getting a `SQLDecodeError` in `_token2op` with the changes. It fails some authentication requests and the app is stuck at login.

**Comment by marqueurs404** [comment]
I've made the changes for the refactoring suggestions. I've also made a separate branch with only the changes for the agg(distinct) fix at https://github.com/morbitech1/djongo/tree/agg-distinct-fix, in case you might wanna merge that first. 

I could help you to take a look at why the auth requests are failing tomorrow.

**Pull request closed without merging**